File: 05_Development/CareBridgeAITriageService/app/context/target_entity_resolver.py
"""Deterministic target-entity resolution: is this session about the mother or the baby?

Getting this wrong is not a UX bug. If the engine decides "mother" for a message about an
infant, it will apply maternal thresholds to a baby and ask the wrong questions. So the
resolver is deliberately conservative: it returns UNKNOWN rather than guess, and CONFLICTED
rather than pick a side.

Precedence (highest first), matching ``ResolutionSource``:

1. an explicit clarification answer in this session
2. an explicit subject in the latest message
3. an explicitly selected profile in the request
4. a target the user already confirmed earlier in the conversation
5. strong stage-specific context
6. extractor inference (Phase 4; not wired yet)

A stored profile never overrides what the user just typed — hence 2 above 3.
"""

from __future__ import annotations

import json
import re
import unicodedata
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from typing import Mapping, Sequence

from app.context.enums import CareStage, ResolutionSource, TargetEntity

INDICATORS_PATH = Path(__file__).resolve().parents[2] / "data" / "target_entity_indicators_v1.json"


@lru_cache(maxsize=1)
def load_indicators() -> Mapping[str, object]:
    return json.loads(INDICATORS_PATH.read_text(encoding="utf-8"))


@dataclass(frozen=True)
class TargetResolution:
    entity: TargetEntity
    source: ResolutionSource
    #: Surface cues that drove the decision, for the audit trail.
    evidence: tuple[str, ...] = ()
    #: Populated when both entities are named, so the UI can say what it saw.
    conflict_evidence: tuple[str, ...] = ()

    @property
    def is_resolved(self) -> bool:
        return self.entity.is_resolved


def _fold(text: str | None) -> str:
    """Lowercase, strip Vietnamese accents, collapse whitespace."""

    if not text:
        return ""
    lowered = text.lower().replace("đ", "d")
    stripped = "".join(
        char for char in unicodedata.normalize("NFD", lowered)
        if unicodedata.category(char) != "Mn"
    )
    return re.sub(r"\s+", " ", stripped).strip()


def _lower(text: str | None) -> str:
    """Lowercase and collapse whitespace, KEEPING accents."""

    return re.sub(r"\s+", " ", (text or "").lower()).strip()


def _clauses(text: str | None) -> list[str]:
    """Split into clauses on the ACCENTED text, then fold each clause for matching.

    Splitting on folded text is wrong in Vietnamese: the conjunction "còn" folds to "con",
    which is the word for a child. "giúp con tôi" would then be cut in half and the baby
    reference lost. Accents are only safe to drop once the boundaries are already fixed.
    """

    rules = load_indicators()["matchingRules"]
    separators = [sep.lower() for sep in rules["clauseSeparators"] if sep.strip() or sep in ",.;!?"]
    pattern = "|".join(re.escape(sep) for sep in separators)
    lowered = _lower(text)
    if not pattern:
        return [_fold(lowered)] if lowered else []
    parts = [part.strip() for part in re.split(pattern, lowered)]
    return [_fold(part) for part in parts if part.strip()]


def _contains_any(haystack: str, phrases: Sequence[str]) -> list[str]:
    """Whole-word phrase matching.

    Raw substring matching is unsafe here: "tã" (nappy) folds to "ta", which then matches
    inside "tay" (hand) and "tập" (exercise), so "đau cổ tay sau khi tập" scored as a baby
    message. Short Vietnamese tokens make this failure mode easy to hit, so every phrase is
    anchored on both sides.
    """

    found = []
    for phrase in phrases:
        needle = _fold(phrase)
        if not needle:
            continue
        if re.search(rf"(?<![a-z0-9]){re.escape(needle)}(?![a-z0-9])", haystack):
            found.append(phrase)
    return found


def score_message(message: str | None) -> tuple[TargetEntity, tuple[str, ...], tuple[str, ...]]:
    """Score a raw message. Returns ``(entity, evidence, conflict_evidence)``."""

    indicators = load_indicators()
    folded = _fold(message)
    if not folded:
        return TargetEntity.UNKNOWN, (), ()
    clauses = _clauses(message)

    multi = _contains_any(folded, indicators["multiEntityPhrases"]["phrases"])
    if multi:
        # "Tôi và bé đều bị sốt" — never silently pick one. One session, one target.
        return TargetEntity.CONFLICTED, (), tuple(multi)

    third_party = _contains_any(folded, indicators["thirdPartyPhrases"]["phrases"])

    mother_hits: list[str] = []
    baby_hits: list[str] = []

    for clause in clauses:
        baby_possessive = _contains_any(clause, indicators["baby"]["possessiveOrSubject"])
        baby_strong = _contains_any(clause, indicators["baby"]["strongIndicators"])
        mother_strong = _contains_any(clause, indicators["mother"]["strongIndicators"])
        mother_subject = _contains_any(clause, indicators["mother"]["possessiveOrSubject"])

        if baby_possessive or baby_strong:
            # A possessed child in the clause outranks the first-person pronoun that
            # introduced it: "tôi hỏi giúp con tôi, bé bị sốt" is about the baby.
            baby_hits.extend(baby_possessive + baby_strong)
            continue
        if mother_strong:
            mother_hits.extend(mother_strong)
            continue
        if mother_subject and not third_party:
            mother_hits.extend(mother_subject)

    if mother_hits and baby_hits:
        return TargetEntity.CONFLICTED, (), tuple(dict.fromkeys(mother_hits + baby_hits))
    if baby_hits:
        return TargetEntity.BABY, tuple(dict.fromkeys(baby_hits)), ()
    if mother_hits:
        return TargetEntity.MOTHER, tuple(dict.fromkeys(mother_hits)), ()
    return TargetEntity.UNKNOWN, (), ()


def _entity_from_clarification(answer: str | None) -> TargetEntity | None:
    return {
        "CLARIFY_TARGET_MOTHER": TargetEntity.MOTHER,
        "CLARIFY_TARGET_BABY": TargetEntity.BABY,
        "CLARIFY_TARGET_BOTH": TargetEntity.CONFLICTED,
    }.get(answer or "")


def _entity_from_stage(stage: CareStage | None) -> TargetEntity | None:
    if stage is None or not stage.is_resolved:
        return None
    if stage in (CareStage.INFANT_0_12M, CareStage.TODDLER_12_24M):
        return TargetEntity.BABY
    if stage in (CareStage.PRECONCEPTION, CareStage.POSSIBLE_PREGNANCY, CareStage.PREGNANCY):
        return TargetEntity.MOTHER
    # POSTPARTUM_MOTHER names the mother's stage, but a postpartum session is exactly when a
    # newborn question is most likely. It is not strong enough evidence on its own.
    return None


def resolve_target_entity(
    *,
    clarification_answer: str | None = None,
    latest_user_message: str | None = None,
    selected_profile_entity: TargetEntity | None = None,
    confirmed_conversation_target: TargetEntity | None = None,
    stage: CareStage | None = None,
) -> TargetResolution:
    """Resolve the session's target entity, or report why it could not be resolved."""

    clarified = _entity_from_clarification(clarification_answer)
    if clarified is not None:
        return TargetResolution(clarified, ResolutionSource.EXPLICIT_CLARIFICATION_ANSWER,
                                evidence=(clarification_answer,))

    entity, evidence, conflicts = score_message(latest_user_message)
    if entity is TargetEntity.CONFLICTED:
        return TargetResolution(entity, ResolutionSource.EXPLICIT_IN_LATEST_MESSAGE,
                                conflict_evidence=conflicts)
    if entity.is_resolved:
        return TargetResolution(entity, ResolutionSource.EXPLICIT_IN_LATEST_MESSAGE, evidence)

    if selected_profile_entity is not None and selected_profile_entity.is_resolved:
        return TargetResolution(selected_profile_entity, ResolutionSource.EXPLICIT_SELECTED_PROFILE)

    if confirmed_conversation_target is not None and confirmed_conversation_target.is_resolved:
        return TargetResolution(confirmed_conversation_target,
                                ResolutionSource.CONFIRMED_CONVERSATION_TARGET)

    from_stage = _entity_from_stage(stage)
    if from_stage is not None:
        return TargetResolution(from_stage, ResolutionSource.STAGE_SPECIFIC_CONTEXT)

    return TargetResolution(TargetEntity.UNKNOWN, ResolutionSource.NONE)

File: 05_Development/CareBridgeAPI/src/main/java/com/carebridge/backend/triage/rules/TargetEntityResolver.java
package com.carebridge.backend.triage.rules;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.InputStream;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Deterministic target-entity resolution: is this session about the mother or the baby?
 *
 * <p>Getting this wrong is not a UX bug. Deciding "mother" for a message about an infant
 * applies maternal thresholds to a baby and asks the wrong questions. The resolver is
 * therefore conservative: UNKNOWN rather than guess, CONFLICTED rather than pick a side.
 *
 * <p>Behavioural parity with {@code app/context/target_entity_resolver.py}; both read the same
 * {@code target_entity_indicators_v1.json}, because two different phrase lists would mean the
 * two runtimes disagree about who the session is about.
 */
@Component
public class TargetEntityResolver {

    public static final String INDICATORS_RESOURCE = "triage/target_entity_indicators_v1.json";

    private final JsonNode indicators;

    public TargetEntityResolver() {
        this(INDICATORS_RESOURCE);
    }

    TargetEntityResolver(String resourcePath) {
        try (InputStream stream = new ClassPathResource(resourcePath).getInputStream()) {
            this.indicators = new ObjectMapper().readTree(stream);
        } catch (IOException exception) {
            throw new IllegalStateException(
                    "cannot read target entity indicators " + resourcePath, exception);
        }
    }

    /** The resolved entity plus how it was reached, for the audit trail. */
    public record TargetResolution(
            TargetEntity entity,
            ResolutionSource source,
            List<String> evidence,
            List<String> conflictEvidence) {

        public boolean isResolved() {
            return entity.isResolved();
        }
    }

    public TargetResolution resolve(
            String clarificationAnswer,
            String latestUserMessage,
            TargetEntity selectedProfileEntity,
            TargetEntity confirmedConversationTarget,
            CareStage stage) {

        TargetEntity clarified = fromClarification(clarificationAnswer);
        if (clarified != null) {
            return new TargetResolution(clarified, ResolutionSource.EXPLICIT_CLARIFICATION_ANSWER,
                    List.of(clarificationAnswer), List.of());
        }

        TargetResolution scored = scoreMessage(latestUserMessage);
        if (scored.entity() == TargetEntity.CONFLICTED || scored.entity().isResolved()) {
            return scored;
        }

        if (selectedProfileEntity != null && selectedProfileEntity.isResolved()) {
            return new TargetResolution(selectedProfileEntity,
                    ResolutionSource.EXPLICIT_SELECTED_PROFILE, List.of(), List.of());
        }
        if (confirmedConversationTarget != null && confirmedConversationTarget.isResolved()) {
            return new TargetResolution(confirmedConversationTarget,
                    ResolutionSource.CONFIRMED_CONVERSATION_TARGET, List.of(), List.of());
        }

        TargetEntity fromStage = fromStage(stage);
        if (fromStage != null) {
            return new TargetResolution(fromStage, ResolutionSource.STAGE_SPECIFIC_CONTEXT,
                    List.of(), List.of());
        }
        return new TargetResolution(TargetEntity.UNKNOWN, ResolutionSource.NONE,
                List.of(), List.of());
    }

    /** Score a raw message alone, with no profile or conversation context. */
    public TargetResolution scoreMessage(String message) {
        String folded = fold(message);
        if (folded.isEmpty()) {
            return new TargetResolution(TargetEntity.UNKNOWN, ResolutionSource.NONE,
                    List.of(), List.of());
        }

        List<String> multi = containsAny(folded,
                indicators.get("multiEntityPhrases").get("phrases"));
        if (!multi.isEmpty()) {
            // "Tôi và bé đều bị sốt" — never silently pick one. One session, one target.
            return new TargetResolution(TargetEntity.CONFLICTED,
                    ResolutionSource.EXPLICIT_IN_LATEST_MESSAGE, List.of(), multi);
        }

        boolean thirdParty = !containsAny(folded,
                indicators.get("thirdPartyPhrases").get("phrases")).isEmpty();

        LinkedHashSet<String> motherHits = new LinkedHashSet<>();
        LinkedHashSet<String> babyHits = new LinkedHashSet<>();

        for (String clause : clauses(message)) {
            List<String> babyPossessive = containsAny(clause,
                    indicators.get("baby").get("possessiveOrSubject"));
            List<String> babyStrong = containsAny(clause,
                    indicators.get("baby").get("strongIndicators"));
            if (!babyPossessive.isEmpty() || !babyStrong.isEmpty()) {
                // A possessed child outranks the pronoun that introduced it:
                // "tôi hỏi giúp con tôi, bé bị sốt" is about the baby.
                babyHits.addAll(babyPossessive);
                babyHits.addAll(babyStrong);
                continue;
            }
            List<String> motherStrong = containsAny(clause,
                    indicators.get("mother").get("strongIndicators"));
            if (!motherStrong.isEmpty()) {
                motherHits.addAll(motherStrong);
                continue;
            }
            List<String> motherSubject = containsAny(clause,
                    indicators.get("mother").get("possessiveOrSubject"));
            if (!motherSubject.isEmpty() && !thirdParty) {
                motherHits.addAll(motherSubject);
            }
        }

        if (!motherHits.isEmpty() && !babyHits.isEmpty()) {
            List<String> conflict = new ArrayList<>(motherHits);
            conflict.addAll(babyHits);
            return new TargetResolution(TargetEntity.CONFLICTED,
                    ResolutionSource.EXPLICIT_IN_LATEST_MESSAGE, List.of(), List.copyOf(conflict));
        }
        if (!babyHits.isEmpty()) {
            return new TargetResolution(TargetEntity.BABY,
                    ResolutionSource.EXPLICIT_IN_LATEST_MESSAGE, List.copyOf(babyHits), List.of());
        }
        if (!motherHits.isEmpty()) {
            return new TargetResolution(TargetEntity.MOTHER,
                    ResolutionSource.EXPLICIT_IN_LATEST_MESSAGE, List.copyOf(motherHits), List.of());
        }
        return new TargetResolution(TargetEntity.UNKNOWN, ResolutionSource.NONE,
                List.of(), List.of());
    }

    private static TargetEntity fromClarification(String answer) {
        if (answer == null) {
            return null;
        }
        return switch (answer) {
            case "CLARIFY_TARGET_MOTHER" -> TargetEntity.MOTHER;
            case "CLARIFY_TARGET_BABY" -> TargetEntity.BABY;
            case "CLARIFY_TARGET_BOTH" -> TargetEntity.CONFLICTED;
            default -> null;
        };
    }

    private static TargetEntity fromStage(CareStage stage) {
        if (stage == null || !stage.isResolved()) {
            return null;
        }
        return switch (stage) {
            case INFANT_0_12M, TODDLER_12_24M -> TargetEntity.BABY;
            case PRECONCEPTION, POSSIBLE_PREGNANCY, PREGNANCY -> TargetEntity.MOTHER;
            // POSTPARTUM_MOTHER names the mother's stage, but postpartum is exactly when a
            // newborn question is most likely. Not strong enough evidence on its own.
            default -> null;
        };
    }

    /**
     * Split into clauses on the ACCENTED text, then fold each clause for matching.
     *
     * <p>Splitting on folded text is wrong in Vietnamese: the conjunction "còn" folds to "con",
     * the word for a child, so "giúp con tôi" would be cut in half and the baby reference lost.
     * Accents are only safe to drop once the boundaries are fixed.
     */
    private List<String> clauses(String message) {
        String lowered = message == null ? "" : message.toLowerCase(Locale.ROOT)
                .replaceAll("\\s+", " ").trim();
        if (lowered.isEmpty()) {
            return List.of();
        }
        List<String> separators = new ArrayList<>();
        indicators.get("matchingRules").get("clauseSeparators")
                .forEach(node -> separators.add(java.util.regex.Pattern.quote(
                        node.asText().toLowerCase(Locale.ROOT))));
        List<String> result = new ArrayList<>();
        for (String part : lowered.split(String.join("|", separators))) {
            String folded = fold(part);
            if (!folded.isEmpty()) {
                result.add(folded);
            }
        }
        return result;
    }

    /**
     * Whole-word phrase matching.
     *
     * <p>Raw substring matching is unsafe here: "tã" (nappy) folds to "ta", which then matches
     * inside "tay" (hand) and "tập" (exercise), so "đau cổ tay sau khi tập" scored as a baby
     * message. Short Vietnamese tokens make this easy to hit, so phrases are anchored.
     */
    static List<String> containsAny(String haystack, JsonNode phrases) {
        List<String> hits = new ArrayList<>();
        for (JsonNode phrase : phrases) {
            String folded = fold(phrase.asText());
            if (folded.isEmpty()) {
                continue;
            }
            java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(
                    "(?<![a-z0-9])" + java.util.regex.Pattern.quote(folded) + "(?![a-z0-9])");
            if (pattern.matcher(haystack).find()) {
                hits.add(phrase.asText());
            }
        }
        return hits;
    }

    /** Lowercase, strip Vietnamese accents, collapse whitespace. */
    static String fold(String text) {
        if (text == null) {
            return "";
        }
        String lowered = text.toLowerCase(Locale.ROOT).replace('đ', 'd');
        return Normalizer.normalize(lowered, Normalizer.Form.NFD)
                .replaceAll("\\p{M}+", "")
                .replaceAll("\\s+", " ")
                .trim();
    }

    /** Exposed for the parity test so both runtimes assert the same source file. */
    Map<String, JsonNode> sections() {
        return Map.of(
                "mother", indicators.get("mother"),
                "baby", indicators.get("baby"),
                "multiEntityPhrases", indicators.get("multiEntityPhrases"),
                "thirdPartyPhrases", indicators.get("thirdPartyPhrases"));
    }
}
